### PR TITLE
Remove warning on `no_backward_sync` with XLA strategy

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes https://github.com/Lightning-AI/lit-parrot/pull/110#discussion_r1218342602

cc @carmocca @justusschock @awaelchli @JackCaoG @steventk-g @Liyang90

### CODE REVIEW

File: src/lightning/fabric/fabric.py [hunk 1]
```diff
@@ -565,14 +565,11 @@ def no_backward_sync(self, module: _FabricModule, enabled: bool = True) -> Gener
...
                 " `model = self.setup(model, ...)`"
             )
-        if not enabled or isinstance(self._strategy, SingleDeviceStrategy):
-            context = nullcontext()
-        elif self._strategy._backward_sync_control is None:
-            rank_zero_warn(
-                f"The `{self._strategy.__class__.__name__}` does not support skipping the gradient synchronization."
-                f" Remove `.no_backward_sync()` from your code or choose a different strategy.",
-                category=PossibleUserWarning,
-            )
```
Comment (awaelchli) [inline src/lightning/fabric/fabric.py:575]: > The warning goes against the no-code-changes philosophy when changing strategies.

IMO it does not. The warning is here to give clear feedback about which strategy supports backward sync control. If it was a no-op, it would give a false sense of optimization. Take for example the DP strategy or DeepSpeed strategy. If I did this:

```py
with fabric.no_backward_sync(module, enabled=True):
    ...
```

then we're asking explicitly to skip sync in backward, yet without a warning it will just not follow our instructions (because it's unsupported, the strategy will do it anyway). If we wanted to be strict, we would have made it an error. But in the spirit of "no-code-changes philosophy when changing strategies" we made it a warning instead. So I strongly recommend that we keep the warning. 

From the linked issue I see that this is motivated by XLA usage. For the XLA strategy, it is a false-positive, so removing the warning for the XLA strategy is fine with me and the right thing todo, since sync happens in the optimizer step (this was a bug, we didn't know better).

File: src/lightning/fabric/fabric.py [hunk 2]
```diff
@@ -565,14 +565,11 @@ def no_backward_sync(self, module: _FabricModule, enabled: bool = True) -> Gener
                 "You need to set up the model first before you can call `self.no_backward_sync()`:"
                 " `model = self.setup(model, ...)`"
             )
-        if not enabled or isinstance(self._strategy, SingleDeviceStrategy):
```
Comment (awaelchli) [inline src/lightning/fabric/fabric.py:568]: I propose adding XLA in this condition here (see other comment)

**Review by awaelchli — CHANGES_REQUESTED** [review]
I left a comment

File: src/lightning/fabric/fabric.py (hunk 1 above)
Comment (carmocca) [reply src/lightning/fabric/fabric.py:575]: A lit-parrot user got confused by the same warning on DeepSpeed: https://github.com/Lightning-AI/lit-parrot/issues/62

What you said makes sense for a fabric training loop that is specific to one strategy, but if your script is meant to allow easily changing this, then the warning becomes annoying as it expects that you write the following to avoid it.

```
if isinstance(fabric.strategy, DeepSpeedStrategy):
    context = nullcontext()
else
    context = fabric.no_backward_sync()
with context):
    ...
```
or that you silence the warning.

Our lit-* repos are like this, where the same fabric code supports multiple strategies.

With this in mind, would you still also keep the warning for DeepSpeed, or would you also explicitly disable it for deepspeed?

If you still want to keep it then I would silence the warning on `lit-parrot`

Comment (awaelchli) [reply src/lightning/fabric/fabric.py:575]: For deepspeed it is semi-legit. Gradient accumulation can be configured in the deepspeed config, and then synchronization happens only when necessary. If it is not configured, or not configured properly, then the context manager's warning does make sense. If it is configured, we could probably parse that and avoid the false-positive warning. So I would keep it for DeepSpeed or try to parse the config.

> What you said makes sense for a fabric training loop that is specific to one strategy, but if your script is meant to allow easily changing this, then the warning becomes annoying as it expects that you write the following to avoid it.

Personal opinion: this is why we should be using the Trainer. I think if this was the goal to have a fully-configureable script for lit-llama and lit-parrot, Trainer should have been chosen in the first place and not Fabric. Fabric will always have these opt-in optional features to get the best out of a specific strategy or specific setup. That's the Fabric lifestyle. Most things are agnosic, but if the user wants, they can reach deeper at the expense of not being as general.

If there is a desire to have a context manager that has this functionality without warnings, then it should be redesigned/renamed to be as neutral as possible in what it expresses. 

On the other hand, if for some reason the warnings are disturbing to users so much, I suggest removing the context managers from the scripts entirely or suppress the warning.

Comment (carmocca) [reply src/lightning/fabric/fabric.py]: Okay. I'll silence it there

**Review by awaelchli — APPROVED** [review]
Great, thanks!

**Comment by awaelchli** [comment]
This could go as a bugfix, as the warning was a false-positive by mistake.

**Review by justusschock — APPROVED** [review]
